python/ray/air/execution/resources/placement_group.py

```diff
@@ -0,0 +1,213 @@
+                    num_cpus=num_cpus,
+                    num_gpus=num_gpus,
+                    memory=memory,
+                    resources=bundle,
+                )
+            )
+        return annotated
+
+
+@DeveloperAPI
+class PlacementGroupResourceManager(ResourceManager):
+    """Resource manager using placement groups as the resource backend.
+
+    Ray core does not emit events when resources are available. Instead, we
+    have to periodically request and update the scheduling state of the
+    placement groups.
+
+    Internally, the placement group lifecycle is like this:
+    - Resources are requested with ``request_resources()``
+    - A placement group is scheduled ("staged")
+    - A ``PlacementGroup.ready()`` is scheduled ("staging future")
+    - We update the scheduling state when we need to
+        (e.g. when ``has_resources_ready()`` is called)
+    - When staging futures resolved, a placement group is moved from "staging"
+        to "ready"
+    - When a resource request is canceled, we remove a placement group from "staging".
+        If there are not staged placement groups (because they are already "ready"),
+        we remove one from "ready" instead.
+    - When a resource is acquired, the pg is removed from "ready" and moved
+        to "acquired"
+    - When a resource is freed, the pg is removed from "acquired" and destroyed
+
+    Per default, placement group scheduling state is refreshed every time when
+    resource state is inquired, but not more often than once every ``update_interval``
+    seconds. Alternatively, staging futures can be retrieved (and awaited) with
+    ``get_resource_futures()`` and state update can be force with ``update_state()``.
+
+    Args:
+        update_interval: Minimum interval in seconds between updating scheduling
+            state of placement groups.
+
+    """
+
+    _resource_cls: AllocatedResource = PlacementGroupAllocatedResource
+
+    def __init__(self, update_interval: float = 0.1):
+        self._pg_to_request: Dict[PlacementGroup, ResourceRequest] = {}
+        self._request_to_staged_pgs: Dict[
+            ResourceRequest, Set[PlacementGroup]
+        ] = defaultdict(set)
+        self._request_to_ready_pgs: Dict[
+            ResourceRequest, Set[PlacementGroup]
+        ] = defaultdict(set)
+
+        self._staging_future_to_pg: Dict[ray.ObjectRef, PlacementGroup] = dict()
+        self._pg_to_staging_future: Dict[PlacementGroup, ray.ObjectRef] = dict()
+        self._acquired_pgs: Set[PlacementGroup] = set()
+
+        self._update_interval = update_interval
+        self._last_update = time.monotonic() - self._update_interval - 1
+
+    def get_resource_futures(self) -> List[ray.ObjectRef]:
+        return list(self._staging_future_to_pg.keys())
+
+    def _maybe_update_state(self):
+        now = time.monotonic()
+        if now > self._last_update + self._update_interval:
+            self.update_state()
+
+    def update_state(self):
```

Comment:
Can we clarify what state is being updated? At least now, this is specific to the asynchronous waiting of placement groups to be scheduled, while other state is all updated synchronously.

Reply:
Is the docstring of the base interface clear enough?

Reply:
To be completely honest, not really 😅 

```
        """Reconcile internal state with cluster state.
        Depending on the backend, resource futures can be awaited externally
        (with ``get_resource_futures()``). If such a future resolved, the outer
        control loop can instruct the resource manager to update its internal state.
        """
```

From this particular comment, I'm wondering why the internal state isn't automatically updated when `get_resource_futures()` is called. More generally, when is state not expected to be already up to date?

Reply:
Ray core does not emit events when resources (e.g. placement groups) are scheduled (or for anything else, really). Instead you have to await scheduling futures (we call them "staging futures" in the placement group resource manager) with `ray.wait()`.

So what we do is when someone asks `has_resources_ready()` we don't always know, yet - we have to check if the staging future resolved. This is what happens in `update_state()`.

Now, we actually do call `update_state()` first in `has_resources_ready()` so it does happen automatically. But we call this function potentially very often in Ray Tune - up to O(num_trials) times _per trial runner iteration_. And each call to update state will wait on all staging futures - up to O(staging_futures). We thus end up with O(num_running_trials * num_pending_trials) _per iteration_, which can break scalability of Ray Tune if we run with thousands of trials.

To prevent this, we only update every 0.1 seconds.

Now, we also have other futures to await, specifically training futures. In the event loop, if there's nothing else to do, we wait until the next future resolves and then react to it (again, because Ray core does not have an event/callback based system). This call to `ray.wait()` has to include all futures - training futures and resource futures. That's why we have `get_resource_futures()`. If the next future to resolve happens to be a resource future, we want the resource manager to immediately update its state so we can schedule the next trial on it. That's why we call `update_state()`.

The two alternatives here are:
1. Remove the 0.1 second break. As explained above this can lead to problems with large experiments
2. Accept a 0.1 delay when resource futures resolve. This is not ideal (what's the point of awaiting the futures in the first place), and also breaks down when we have a lot of very short running trials (where a 0.1 delay per trial adds up quite a bit).

The `get_resource_futures()/update_state()` combination is not ideal, but it's a good tradeoff to achieve what we want without leaking too much responsibility to the caller. After all, the caller _does not have to_ use this. It's just an optimization to enable external awaiting of resource futures. The caller can also just periodically ask if resources are ready and will see them ready at most 0.1 seconds later than without using this.